src/OpenSage.Game/DataStructures/QuadtreeNode.cs

```diff
@@ -0,0 +1,196 @@
+﻿using System.Collections.Generic;
+using System.Numerics;
+using OpenSage.Logic.Object;
+using OpenSage.Mathematics;
+
+namespace OpenSage.DataStructures;
+
+public class QuadtreeNode<T> where T : class, ICollidable
+{
+    private readonly RectangleF _box;
+    private readonly int _maxItemsPerNode;
+    private readonly HashSet<T> _data;
+    private readonly QuadtreeNode<T>[] _children = new QuadtreeNode<T>[4];
+    private bool _hasChildren;
+    private readonly RectangleF _originalSize;
+    private readonly float _overlapFactor;
+    private readonly float _minNodeSideLength;
+    private readonly Vector2 _center;
+
+    private bool IsMinSize => _originalSize.Width * _overlapFactor / 2 < _minNodeSideLength || _originalSize.Height * _overlapFactor / 2 < _minNodeSideLength;
+
+    public QuadtreeNode(in RectangleF bounds, in RectangleF originalSize, float minNodeSideLength, float scaleFactor, int maxItemsPerNode)
+    {
+        _box = bounds;
+        _overlapFactor = scaleFactor;
+        _originalSize = originalSize;
+        _minNodeSideLength = minNodeSideLength;
+        _maxItemsPerNode = maxItemsPerNode;
+        _center = new Vector2(bounds.X + bounds.Width / 2, bounds.Y + bounds.Height / 2);
+        _data = new HashSet<T>(maxItemsPerNode);
+    }
+
+    public IEnumerable<T> Find(Collider collider, T? searcher, bool twoDimensional, bool hasToCollideWithSearcher)
+    {
+        if (_hasChildren)
+        {
+            foreach (var child in _children)
+            {
+                if (!child.Touches(collider))
+                {
+                    continue;
+                }
+
+                foreach (var subChild in child.Find(collider, searcher, twoDimensional, hasToCollideWithSearcher))
+                {
+                    if (CollidesForReal(subChild, collider, searcher, twoDimensional, hasToCollideWithSearcher))
+                    {
+                        yield return subChild;
+                    }
+                }
+            }
+        }
+
+        foreach (var data in _data)
+        {
+            if (!collider.Intersects(data.RoughCollider.AxisAlignedBoundingArea))
+            {
+                continue;
+            }
+
+            if (CollidesForReal(data, collider, searcher, twoDimensional, hasToCollideWithSearcher))
+            {
+                yield return data;
+            }
+        }
+    }
+
+    private static bool CollidesForReal(T item, Collider collider, T? searcher, bool twoDimensional, bool hasToCollideWithSearcher)
```

Comment:
I just realized I never went back and renamed this - open to suggestions for a better method name here 😆

Reply:
Can you describe in one sentence what the method does? I've read the code but I'm not 100% sure - is it a precise collision test?

Reply:
Ah, that's it - `CollidesPrecise` should do the trick.

Yeah, this was ported over from the original implementation, but basically the quadtree does most of its checks on the GameObject's `RoughCollider` (so, an AABB), and then if that matches then this precise version is called which checks all the _actual_ collidables against each other.